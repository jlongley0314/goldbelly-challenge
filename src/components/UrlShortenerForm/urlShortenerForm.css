.ShortenUrlTitle {
  text-align: center;
  padding: 20px;
  border-bottom: 1px solid rgba(0, 0, 0, 0.125);
}
