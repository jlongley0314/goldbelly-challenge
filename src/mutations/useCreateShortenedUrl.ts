import { useMutation } from "react-query";

export type UrlShortenerFormData = {
  url: string;
  short_url: string;
  slug?: string;
};

export function useCreateShortenedUrl(accessToken: string) {
  return useMutation(async (formData: UrlShortenerFormData) => {
    let headers = new Headers();
    headers.append("GB-Access-Token", accessToken);
    headers.append("Content-Type", "application/json");
    const body = JSON.stringify(formData);
    const requestOptions = {
      method: "POST",
      headers: headers,
      body: body,
    };
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/links`,
      requestOptions
    );

    if (!response.ok) {
      const error = await response.json();
      if (error.errors && Object.keys(error.errors)[0]) {
        const errorKey = Object.keys(error.errors)[0];
        const errorValue = Object.values(error.errors)[0];
        const newError = new Error();
        newError.message = `${errorKey} ${errorValue}`;
        throw newError;
      } else {
        throw new Error("Something went wrong.");
      }
    }

    return response;
  });
}
