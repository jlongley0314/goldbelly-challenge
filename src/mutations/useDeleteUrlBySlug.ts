import { useMutation } from "react-query";

export function useDeleteUrlBySlug(accessToken: string) {
  return useMutation(async (slug: string) => {
    let headers = new Headers();
    headers.append("GB-Access-Token", accessToken);
    headers.append("Content-Type", "application/json");
    const requestOptions = {
      method: "DELETE",
      headers: headers,
    };
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/links/${slug}`,
      requestOptions
    );

    if (!response.ok) {
      const error = await response.json();
      throw error.error;
    }

    return response;
  });
}
