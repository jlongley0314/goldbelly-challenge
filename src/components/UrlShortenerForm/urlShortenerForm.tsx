import "./urlShortenerForm.css";
import { useState } from "react";
import Card from "react-bootstrap/Card";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import { useCreateShortenedUrl } from "../../mutations/useCreateShortenedUrl";

export function UrlShortenerForm() {
  const [validated, setValidated] = useState(false);
  const {
    mutate: createShortenedUrl,
    isSuccess: createShortenedUrlSuccess,
    isError: createShortenedUrlError,
  } = useCreateShortenedUrl(process.env.REACT_APP_API_KEY ?? "");
  const handleSubmit = (event: {
    currentTarget: any;
    preventDefault: () => void;
    stopPropagation: () => void;
  }) => {
    const form = event.currentTarget;
    if (form.checkValidity() === false) {
      event.preventDefault();
      event.stopPropagation();
    }

    setValidated(true);
  };
  return (
    <Card>
      <Card.Title className="ShortenUrlTitle">Shorten Your URL</Card.Title>
      <Card.Body>
        <Form validated={validated} onSubmit={handleSubmit}>
          <Form.Group className="mb-3">
            <Form.Label>Long URL</Form.Label>
            <Form.Control
              data-testid="LongUrlInput"
              required
              type="text"
              placeholder="Long URL"
            />
          </Form.Group>
          <Form.Group className="mb-3">
            <Form.Label>Short URL</Form.Label>
            <Form.Control
              data-testid="ShortUrlInput"
              required
              type="text"
              placeholder="Short URL"
            />
          </Form.Group>
          <Form.Group className="mb-3">
            <Form.Label>Slug</Form.Label>
            <Form.Control
              data-testid="SlugInput"
              type="text"
              placeholder="Slug (optional)"
            />
          </Form.Group>
          <div className="d-grid">
            <Button type="submit">Submit</Button>
          </div>
        </Form>
      </Card.Body>
    </Card>
  );
}
