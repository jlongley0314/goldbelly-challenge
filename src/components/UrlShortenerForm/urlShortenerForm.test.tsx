import {
  render,
  RenderResult,
  fireEvent,
  screen,
} from "@testing-library/react";
import { UrlShortenerForm } from "./urlShortenerForm";
import { useCreateShortenedUrl } from "../../mutations/useCreateShortenedUrl";
jest.mock("../../mutations/useCreateShortenedUrl");
import { useQueryClient } from "react-query";
jest.mock("react-query");

describe("URLShortenerForm", () => {
  describe("URLShortenerForm success", () => {
    let component: RenderResult;
    let mutate = jest.fn();
    let invalidateQueries = jest.fn();
    beforeEach(() => {
      (useCreateShortenedUrl as jest.Mock).mockReturnValue({
        isLoading: false,
        isError: false,
        error: undefined,
        isSuccess: true,
        mutate: mutate,
      });
      (useQueryClient as jest.Mock).mockReturnValue({ invalidateQueries });
      component = render(<UrlShortenerForm />);
    });

    it("should render component as expected", () => {
      expect(component).toMatchSnapshot();
    });

    it("does not call query when long url is not entered", () => {
      fireEvent.change(screen.getByTestId("LongUrlInput"), {
        target: { value: "test-long" },
      });
      fireEvent.click(screen.getByText("Submit"));
      expect(mutate).not.toHaveBeenCalled();
    });

    it("does not call query when short url is not entered", () => {
      fireEvent.change(screen.getByTestId("ShortUrlInput"), {
        target: { value: "test-short" },
      });
      fireEvent.click(screen.getByText("Submit"));
      expect(mutate).not.toHaveBeenCalled();
    });

    it("should call mutation when form is filled out with correct values", () => {
      fireEvent.change(screen.getByTestId("LongUrlInput"), {
        target: { value: "test-long" },
      });
      fireEvent.change(screen.getByTestId("ShortUrlInput"), {
        target: { value: "test-short" },
      });
      fireEvent.change(screen.getByTestId("SlugInput"), {
        target: { value: "test-slug" },
      });
      fireEvent.click(screen.getByText("Submit"));
      expect(mutate).toHaveBeenCalledWith({
        short_url: "test-short",
        slug: "test-slug",
        url: "test-long",
      });
      expect(invalidateQueries).toHaveBeenCalledWith("getAllGeneratedUrls");
    });
  });

  describe("URLShortenerForm success", () => {
    let mutate = jest.fn();
    let invalidateQueries = jest.fn();
    beforeEach(() => {
      (useCreateShortenedUrl as jest.Mock).mockReturnValue({
        isLoading: false,
        isError: true,
        error: undefined,
        isSuccess: false,
        mutate: mutate,
      });
      (useQueryClient as jest.Mock).mockReturnValue({ invalidateQueries });
      render(<UrlShortenerForm />);
    });

    it("should render the error toast when isError is true", () => {
      expect(screen.getByText("Error Creating Short Url")).toBeInTheDocument();
    });
  });
});
