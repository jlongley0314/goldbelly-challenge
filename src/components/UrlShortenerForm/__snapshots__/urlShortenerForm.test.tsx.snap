// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`URLShortenerForm URLShortenerForm success should render component as expected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="ShortenUrlTitle card-title h5"
        >
          Shorten Your URL
        </div>
        <div
          class="card-body"
        >
          <form
            class=""
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
              >
                Long URL
              </label>
              <input
                class="form-control"
                data-testid="LongUrlInput"
                placeholder="Long URL"
                required=""
                type="text"
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
              >
                Short URL
              </label>
              <input
                class="form-control"
                data-testid="ShortUrlInput"
                placeholder="Short URL"
                required=""
                type="text"
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
              >
                Slug
              </label>
              <input
                class="form-control"
                data-testid="SlugInput"
                placeholder="Slug (optional)"
                type="text"
              />
            </div>
            <div
              class="d-grid"
            >
              <button
                class="btn btn-primary"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
    >
      <div
        class="ShortenUrlTitle card-title h5"
      >
        Shorten Your URL
      </div>
      <div
        class="card-body"
      >
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              Long URL
            </label>
            <input
              class="form-control"
              data-testid="LongUrlInput"
              placeholder="Long URL"
              required=""
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              Short URL
            </label>
            <input
              class="form-control"
              data-testid="ShortUrlInput"
              placeholder="Short URL"
              required=""
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              Slug
            </label>
            <input
              class="form-control"
              data-testid="SlugInput"
              placeholder="Slug (optional)"
              type="text"
            />
          </div>
          <div
            class="d-grid"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
